Babbage / Conway PParams JSON serialization

ch1bo opened the pull request:
# Description

This adds JSON deserialization for `Babbage` and `Conway` `PParams` and is an extension of https://github.com/IntersectMBO/cardano-ledger/pull/3949.

* Creates `Test.Cardano.Ledger.Core.JSON` with a `ToJSON -> FromJSON` roundtrip

* Fixes the `ToJSON` instance of `BabbagePParams` which was missing a `protocolVersion`

* Adds JSON roundtrip tests to Alonzo, Babbage and Conway PParams.

* Adds `protocolVersion` also to `Conway` `PParams` serialization.

* Updated the golden `genesis-conway.json` as `FromJSON` instances are non-derived now and match the `ToJSON` instances for `PoolVotingThresholds` and `DRepVotingThresholds`

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

ch1bo committed:
Create a JSON roundtrip test for Babbage PParams

This will also be useful to test other JSON roundtrips.

ch1bo committed:
Fix Babbage PParams JSON roundtrip

This also adds a JSON PParams roundtrip test for Alonzo.

ch1bo committed:
Add Conway PParams roundtrip and fix JSON instances

When adding a FromJSON for Conway PParams and testing the roundtrip, it
turned out that the JSON instances were not aligned correctly.

ch1bo commented on eras/conway/impl/src/Cardano/Ledger/Conway/PParams.hs:
```diff
@@ -727,6 +769,8 @@ instance Crypto c => ToJSON (ConwayPParams Identity (ConwayEra c)) where
   toEncoding = pairs . mconcat . conwayPParamsPairs
 
+-- TODO: Add protocolVersion to JSON output (of PParams, NOT PParamsUpdate) with
+-- something like: [("protocolVersion", toJSON $ pp ^. ppProtocolVersionL)]
```
@lehins I think I want to add a `protolVersion` entry here, but fail to satisfy constraints because `ProtVerAtMost` on `hkdProtocolVersionL` and couldn't make `ppProtocolVersionL` work. Can you help here?

ch1bo committed:
Add protocolVersion to Conway PParams JSON

This fixes most of the roundtrips for Conway PParams, but seemingly
something changed on the cost models still.

lehins commented on libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/JSON.hs:
```diff
@@ -0,0 +1,38 @@
+  prop (show (typeRep $ Proxy @t)) $ roundTripProperty @t
+
+-- | Roundtrip JSON testing for types and type families that implement
+-- ToJSON/FromJSON. Requires TypeApplication of an @@era@
```
```suggestion
-- ToJSON/FromJSON.
```

lehins reviewed (commented):
I think it would be nice to add to Babbage changelog that protocolVersion issue is fixed. 

Two issues on CI:

# Golden test for ConwayGenesis

The `conway-genesis.json` need to change, since `FromJSON` was wrongfully derived for threasholds and you fixed it in the code (thank you!):
```diff
{
  "poolVotingThresholds": {
-    "pvtCommitteeNormal": 0,
-    "pvtCommitteeNoConfidence": 0,
-    "pvtHardForkInitiation": 0,
-    "pvtMotionNoConfidence": 0
+    "committeeNormal": 0,
+    "committeeNoConfidence": 0,
+    "hardForkInitiation": 0,
+    "motionNoConfidence": 0
  },
  "dRepVotingThresholds": {
-    "dvtMotionNoConfidence": 0,
-    "dvtCommitteeNormal": 0,
-    "dvtCommitteeNoConfidence": 0,
-    "dvtUpdateToConstitution": 0,
-    "dvtHardForkInitiation": 0,
-    "dvtPPNetworkGroup": 0,
-    "dvtPPEconomicGroup": 0,
-    "dvtPPTechnicalGroup": 0,
-    "dvtPPGovGroup": 0,
-    "dvtTreasuryWithdrawal": 0
+    "motionNoConfidence": 0,
+    "committeeNormal": 0,
+    "committeeNoConfidence": 0,
+    "updateToConstitution": 0,
+    "hardForkInitiation": 0,
+    "ppNetworkGroup": 0,
+    "ppEconomicGroup": 0,
+    "ppTechnicalGroup": 0,
+    "ppGovGroup": 0,
+    "treasuryWithdrawal": 0
  },
```

# CostModels in PParams

We need to adjust how we deal with problematic CostModels, since currently they do not roundtrip. I have an idea on what we can do about this, so I'll submit a separate PR that fixes that.

ch1bo committed:
Change golden conway-genesis.json

This was using the field names from the haskell type and are fixed now
to be roundtripping without the prefixes.

ch1bo committed:
Update libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Core/JSON.hs

Co-authored-by: Alexey Kuleshevich <alexey.kuleshevich@iohk.io>

ch1bo committed:
Add changelog entries

ch1bo committed:
Disable JSON roundtrip tests for Conway PParams

The serialization of included cost models is problematic.

ch1bo commented:
@lehins Fixed the golden `genesis-conway` file and added changelog entries. I also disabled the Conway PParams JSON roundtrip tests as I do understand that you will fix the that in another PR.

lehins reviewed (approved):
That's perfect. Thank you!
